include/RAJA/policy/atomic_builtin.hpp

```diff
@@ -37,389 +44,1368 @@ namespace RAJA
+         RAJA::util::reinterp_A_as_B<T, long>(b);
 }
 
+template <typename T,
+          std::enable_if_t<!std::is_same<T, long long>::value &&
+                           sizeof(T) == sizeof(long long), bool> = true>
+RAJA_DEVICE_HIP
+RAJA_INLINE bool builtin_atomicCAS_equal(const T &a, const T &b)
+{
+  return RAJA::util::reinterp_A_as_B<T, long long>(a) ==
+         RAJA::util::reinterp_A_as_B<T, long long>(b);
+}
+
+
+#else  // RAJA_COMPILER_MSVC
+
+
+/*!
+ * Atomic load, store, exchange, and compare and swap
+ */
+template <typename T,
+          std::enable_if_t<(std::is_integral<T>::value ||
+                            std::is_enum<T>::value) &&
+                           (sizeof(T) == 1 ||
+                            sizeof(T) == 2 ||
+                            sizeof(T) == 4 ||
+                            sizeof(T) == 8), bool> = true>
 RAJA_DEVICE_HIP
-RAJA_INLINE unsigned long long builtin_atomic_load(
-    unsigned long long *acc)
+RAJA_INLINE T builtin_atomicLoad(T *acc)
 {
   return __atomic_load_n(acc, __ATOMIC_RELAXED);
 }
 
+template <typename T,
+          std::enable_if_t<(std::is_integral<T>::value ||
+                            std::is_enum<T>::value) &&
+                           (sizeof(T) == 1 ||
+                            sizeof(T) == 2 ||
+                            sizeof(T) == 4 ||
+                            sizeof(T) == 8), bool> = true>
 RAJA_DEVICE_HIP
-RAJA_INLINE void builtin_atomic_store(unsigned *acc,
-                                      unsigned value)
+RAJA_INLINE void builtin_atomicStore(T *acc, T value)
 {
   __atomic_store_n(acc, value, __ATOMIC_RELAXED);
 }
 
+template <typename T,
+          std::enable_if_t<(std::is_integral<T>::value ||
+                            std::is_enum<T>::value) &&
+                           (sizeof(T) == 1 ||
+                            sizeof(T) == 2 ||
+                            sizeof(T) == 4 ||
+                            sizeof(T) == 8), bool> = true>
 RAJA_DEVICE_HIP
-RAJA_INLINE void builtin_atomic_store(
-    unsigned long long *acc,
-    unsigned long long value)
+RAJA_INLINE T builtin_atomicExchange(T *acc, T value)
 {
-  __atomic_store_n(acc, value, __ATOMIC_RELAXED);
+  return __atomic_exchange_n(acc, value, __ATOMIC_RELAXED);
 }
 
+template <typename T,
+          std::enable_if_t<(std::is_integral<T>::value ||
+                            std::is_enum<T>::value) &&
+                           (sizeof(T) == 1 ||
+                            sizeof(T) == 2 ||
+                            sizeof(T) == 4 ||
+                            sizeof(T) == 8), bool> = true>
 RAJA_DEVICE_HIP
-RAJA_INLINE unsigned builtin_atomic_CAS(unsigned *acc,
-                                        unsigned compare,
-                                        unsigned value)
+RAJA_INLINE T builtin_atomicCAS(T *acc, T compare, T value)
 {
   __atomic_compare_exchange_n(
       acc, &compare, value, false, __ATOMIC_ACQ_REL, __ATOMIC_RELAXED);
   return compare;
 }
```

Comment:
Shouldn't this be relaxed memorder in both cases.

Reply:
I was wondering about that. I used the same parameters that were there previously, but if you think it is safe, I will change to relaxed.

Reply:
Ya, we made a conscious decision that all raja atomics are relaxed. If you want anything more nuanced that's where desul comes in.